src/kernel/filesystem/fat32.zig

```diff
@@ -0,0 +1,1756 @@
+            };
+
+            return fat32_fs;
+        }
+    };
+}
+///
+/// Initialise a FAT32 filesystem. This will take a stream that conforms to the reader(), writer(),
+/// and seekableStream() interfaces. For example, FixedBufferStream or File. A pointer to this is
+//// allowed. This will allow the use of different devices such as Hard drives or RAM disks to have
+/// a FAT32 filesystem abstraction. We assume the FAT32 will be the only filesystem present and no
+/// partition schemes are present. So will expect a valid boot sector.
+///
+/// Arguments:
+///     IN allocator: *Allocator - An allocator.
+///     IN stream: anytype       - A stream this is used to read and seek a raw disk or memory to
+///                                be parsed as a FAT32 filesystem. E.g. a FixedBufferStream.
+///
+/// Return: *Fat32FS(@TypeOf(stream))
+///     A pointer to a FAT32 filesystem.
+///
+/// Error: Allocator.Error || Fat32FS(@TypeOf(stream)).Error
+///     Allocator.Error - If there isn't enough memory to create the filesystem.
+///
+pub fn initialiseFAT32(allocator: *Allocator, stream: anytype) (Allocator.Error || ErrorSet(@TypeOf(stream)) || Fat32FS(@TypeOf(stream)).Error)!*Fat32FS(@TypeOf(stream)) {
+    return Fat32FS(@TypeOf(stream)).init(allocator, stream);
+}
+
+/// The number of allocations that the init function make.
+const init_allocations: usize = 5;
+
+/// The test buffer for the test filesystem stream.
+var stream_buff: [if (builtin.is_test) 34090496 else 0]u8 = undefined;
+
+///
+/// Read the test files and write them to the test FAT32 filesystem.
+///
+/// Arguments:
+///     IN comptime StreamType: type     - The stream type.
+///     IN fat32fs: *Fat32FS(StreamType) - The test filesystem.
+///
+/// Error: Allocator.Error || ErrorSet(StreamType) || vfs.Error || std.fs.File.OpenError || std.fs.File.ReadError
+///     Allocator.Error       - Error when allocating memory for reading the file content.
+///     ErrorSet(StreamType)  - Error when writing to the test filesystem.
+///     vfs.Error             - Error with VFS operations.
+///     std.fs.File.OpenError - Error when opening the test files.
+///     std.fs.File.ReadError - Error when reading the test files.
+///
+fn writeTestFiles(comptime StreamType: type, fat32fs: *Fat32FS(StreamType)) (Allocator.Error || ErrorSet(StreamType) || vfs.Error || std.fs.File.OpenError || std.fs.File.ReadError)!void {
+    vfs.setRoot(fat32fs.root_node.node);
+
+    const test_path_prefix = "test/fat32";
+    const files = &[_][]const u8{
+        "/....leading_dots.txt",
+        "/[nope].txt",
+        "/A_verY_Long_File_namE_With_normal_Extension.tXt",
+        "/dot.in.file.txt",
+        "/file.long_ext",
+        "/file.t x t",
+        "/insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long.txt",
```

Comment:
nice xD

Reply:
If you can think of other test files, let me know